### PR TITLE
[ORA-823][ORA-879][ORA-824][ORA-822][ORA-825] Divide mint emisisons by percentage for reputers+workers, pay rest to cosmos validators

### PR DESCRIPTION
**Pull request by relyt29**
First pass at implementing a division of validator incentives vs reputer+worker incentives

validator incentives paid out by distribution module
reputer+worker incentives paid out by emissions module rewards_calc.go

the ratio of validator percent vs reputer+worker percent of fees collected + new emissions is controlled by a new param called PercentRewardsReputersWorkers and set right now to 50% split between the two groups (happy to change as needed)

### CODE REVIEW

File: x/emissions/module/rewards_calc.go [hunk 1]
```diff
@@ -300,6 +291,8 @@ func emitRewardsToTopicParticipants(
 	for participant, reward := range rewards {
 		fmt.Printf(" Emitting %suallo to %s \n", reward.String(), participant)
 		am.keeper.AddStake(ctx, []uint64{topic}, participant, participant, *reward)
+		rewardCoins := sdk.NewCoins(sdk.NewCoin(params.DefaultBondDenom, cosmosMath.NewIntFromBigInt(reward.BigInt())))
+		am.keeper.BankKeeper().SendCoinsFromModuleToModule(ctx, state.AlloraRewardsAccountName, state.AlloraStakingAccountName, rewardCoins)
```
Comment (guilherme-brandao) [inline x/emissions/module/rewards_calc.go:295]: not sure if I got this right: apparently you're double spending here. Adding stake to the account and sending the same amount to the AlloraStakingAccount

File: x/emissions/keeper/keeper.go [hunk 2]
```diff
@@ -230,6 +232,10 @@ func (k *Keeper) GetParams(ctx context.Context) (state.Params, error) {
 	return ret, nil
 }
 
+func (k *Keeper) FeeCollectorName() string {
+	return k.feeCollectorName
+}
+
```
Comment (guilherme-brandao) [inline x/emissions/keeper/keeper.go:238]: *`GetFeeCollectorName`

File: x/emissions/proto/emissions/state/v1/types.proto [hunk 3]
```diff
@@ -11,37 +11,32 @@ import "gogoproto/gogo.proto";
...
-  uint64 min_request_cadence = 11;             // Fastest allowed cadence of a repeating inference request
-  uint64 min_weight_cadence = 12;              // Fastest allowed cadence of uploading weights
-  uint64 max_inference_request_validity = 13;  // longest time a subscription of inferences is allowed to be valid
-  uint64 max_request_cadence = 14;  // slowest (largest) amount of time a subscription can take between inferences
+  uint64 remove_stake_delay_window = 8;        // how long to wait before allowed to remove stake
+  uint64 min_request_cadence = 9;              // Fastest allowed cadence of a repeating inference request
+  uint64 min_weight_cadence = 10;              // Fastest allowed cadence of uploading weights
+  uint64 max_inference_request_validity = 11;  // longest time a subscription of inferences is allowed to be valid
+  uint64 max_request_cadence = 12;  // slowest (largest) amount of time a subscription can take between inferences
+  uint64 percent_rewards_reputers_workers = 13;  // (tx fees + inflation) paid to workers/reputers. Validators get rest
```
Comment (guilherme-brandao) [inline x/emissions/proto/emissions/state/v1/types.proto:39]: I think it's better to use `cosmos.Dec` to percentages, like the inflation:

```
string percent_rewards_reputers_workers = 13 [
    (cosmos_proto.scalar)  = "cosmos.Dec",
    (gogoproto.customtype) = "cosmossdk.io/math.LegacyDec",
    (gogoproto.nullable)   = false,
    (amino.dont_omitempty) = true
  ];
```

**Review by guilherme-brandao — CHANGES_REQUESTED** [review]
Going well! We will need to make sure that we are front running the distribution module indeed during test

File: x/emissions/module/rewards_calc.go (hunk 1 above)
Comment (relyt29) [reply x/emissions/module/rewards_calc.go:295]: one operation sends the tokens, the other increments the data structures that track their proportion of total stake

Comment (relyt29) [reply x/emissions/module/rewards_calc.go:295]: AddStake doesn't actually move tokens itself

File: x/emissions/module/rewards_calc_internal_test.go [hunk 4]
```diff
@@ -38,7 +38,7 @@ func (s *RewardsCalcTestSuite) SetupTest() {
 	addressCodec := address.NewBech32Codec(params.Bech32PrefixAccAddr)
 
 	s.ctx = ctx
-	s.emissionsKeeper = keeper.NewKeeper(encCfg.Codec, addressCodec, storeService, s.authKeeper, s.bankKeeper)
+	s.emissionsKeeper = keeper.NewKeeper(encCfg.Codec, addressCodec, storeService, s.authKeeper, s.bankKeeper, "fee_collector")
```
[the remark this answers is not in the document]
Comment (relyt29) [reply x/emissions/module/rewards_calc_internal_test.go:41]: I just followed what the mint and other modules do for this, they keep it as a param in the keeper for some reason

I kind of agree that we could just use authtypes.FeeCollectorName directly, I don't know why they do it

I guess doing it that way does allow the people controlling the chain to change the feeCollectorName for a module in the chain config if they wanted to

File: integration/topic.sh [hunk 5]
```diff
@@ -41,7 +37,9 @@ $ALLORAD_BIN tx emissions push-topic \
   "$PT_INFERENCE_METHOD" \
   "$PT_INFERENCE_CADENCE" \
   "$PT_DEFAULT_ARG" \
-  --yes --keyring-backend=test --chain-id=demo;
+  --yes --keyring-backend=test --chain-id=demo \
+  --gas-prices=1uallo --gas=auto --gas-adjustment=1.5;
```
Comment (guilherme-brandao) [inline integration/topic.sh:41]: It's optional but you can remove this since we removed the gas fees again

**Review by dmikey — APPROVED** [review]

File: x/emissions/module/rewards_calc.go [hunk 6]
```diff
@@ -158,12 +159,13 @@ func MaskWeightsIfInsufficientLiveness(
 	weights map[string]map[string]*Uint) (map[string]map[string]*Uint, error) {
 	maskedWeights := make(map[string]map[string]*Uint)
 	for reputer, workerWeights := range weights {
-
+		fmt.Println("Masking: ", reputer, " ", workerWeights)
```
Comment (guilherme-brandao) [inline x/emissions/module/rewards_calc.go:162]: Remove print

File: x/mint/keeper/keeper.go [hunk 7]
```diff
@@ -98,12 +98,14 @@ func (k Keeper) MintCoins(ctx context.Context, newCoins sdk.Coins) error {
...
 	}
 
+	//fmt.Println("Mint module: minting coins ", newCoins, " to ", types.ModuleName)
 	return k.bankKeeper.MintCoins(ctx, types.ModuleName, newCoins)
 }
 
 // AddCollectedFees implements an alias call to the underlying supply keeper's
 // AddCollectedFees to be used in BeginBlocker.
 func (k Keeper) AddCollectedFees(ctx context.Context, fees sdk.Coins) error {
+	//fmt.Println("Mint module: adding collected fees ", fees, " to ", k.feeCollectorName)
```
Comment (guilherme-brandao) [inline x/mint/keeper/keeper.go:108]: Remove or add logs

**Review by guilherme-brandao — DISMISSED** [review]

File: app/app.yaml [hunk 8]
```diff
@@ -6,11 +6,12 @@ modules:
       # During begin block slashing happens after distr.BeginBlocker so that
       # there is nothing left over in the validator fee pool, so as to keep the CanWithdrawInvariant invariant.
       # NOTE: staking module is required if HistoricalEntries param > 0
-      begin_blockers: [distribution, staking, mint]
+      # order of emissions must come before distribution for reputers + workers take their rewards cut before validators cut
+      begin_blockers: [emissions, distribution, staking, mint]
```
[the remark this answers is not in the document]
Comment (relyt29) [reply app/app.yaml:10]: I'm getting rid of it, we don't use it

File: integration/topic.sh (hunk 5 above)
Comment (relyt29) [reply integration/topic.sh:41]: i'm going to keep them, makes the tests less likely to arbitrarily fail is my feeling

**Review by kpeluso — APPROVED** [review]

**Review by guilherme-brandao — APPROVED** [review]

**Pull request merged**